fixes issue #13981: unsafe_writes block appeared too late in the atom…

pgrange opened the pull request:
##### ISSUE TYPE

<!--- Pick one below and delete the rest: -->
- Feature Pull Request
- New Module Pull Request
- Bugfix Pull Request
- Docs Pull Request
##### COMPONENT NAME

<!--- Name of the plugin/module/task -->
##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```

```
##### SUMMARY

<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

<!-- Paste verbatim command output below, e.g. before and after your change -->

```

```

…ic_move

workflow. This led to errno.EBUSY to not be managed in the context of
issue #!#981

Commit:
fixes issue #13981: unsafe_writes block appeared too late in the atomic_move
workflow. This led to errno.EBUSY to not be managed in the context of
issue #!#981

bcoca commented:
not sure what you are trying to fix here, unsafe writes should be 'last resort', not a top option

pgrange commented:
Sorry, I did not format my commit comment right for github.

This is related to issue #13981

You are right, unsafe_write should be last resort. However with current code the unsafe_write block is never executed in the context of issue #13981.

Maybe you might prefer another version of the patch but with current code, when we encounter a problem with atomic move, if it is not one of [errno.EPERM, errno.EXDEV, errno.EACCES, errno.ETXTBSY] then we fail.

What I need is to not fail if it is a errno.EBUSY so that unsafe_writes is managed.

pgrange commented:
I updated this repo to reproduce the corresponding issue : 
https://github.com/pgrange/ansible_docker_resource_busy

Commit:
Reduce changes to fix #13981

pgrange commented:
Hi @bcoca,

I updated this pull request to minimize changes. I think it clarifies the issue by changing just two lines.

Let me know if you prefer closing this pull request so that I open a new cleaner one.

Regards,

bcoca commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1941,7 +1941,7 @@ def atomic_move(self, src, dest, unsafe_writes=False):
         except (IOError, OSError):
             e = get_exception()
-            if e.errno not in [errno.EPERM, errno.EXDEV, errno.EACCES, errno.ETXTBSY]:
+            if e.errno not in [errno.EPERM, errno.EXDEV, errno.EACCES, errno.ETXTBSY, errno.EBUSY]:
```
so this makes more sense now

bcoca commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1993,7 +1993,7 @@ def atomic_move(self, src, dest, unsafe_writes=False):
                                 try:
                                     out_dest = open(b_dest, 'wb')
-                                    in_src = open(b_src, 'rb')
+                                    in_src = open(b_tmp_dest_name, 'rb')
```
are you sure this is what we want? b_src should be what we 'tried to copy to b_tmp_dest_name'

pgrange replied:
Not sure why but at this point in code, b_src refers to a non existing file when b_tmp_dest_name refers to a file that exists and has correct data inside.

When running with b_src we have this error:

```
TASK [insert 127.0.0.2 line in /etc/hosts] *************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: UnboundLocalError: local variable 'in_src' referenced before assignment
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_WvGe4f/ansible_module_lineinfile.py\", line 425, in <module>\n    main()\n  File \"/tmp/ansible_WvGe4f/ansible_module_lineinfile.py\", line 417, in main\n    ins_aft, ins_bef, create, backup, backrefs)\n  File \"/tmp/ansible_WvGe4f/ansible_module_lineinfile.py\", line 298, in present\n    write_changes(module, b_lines, dest)\n  File \"/tmp/ansible_WvGe4f/ansible_module_lineinfile.py\", line 175, in write_changes\n    unsafe_writes=module.params['unsafe_writes'])\n  File \"/tmp/ansible_WvGe4f/ansible_modlib.zip/ansible/module_utils/basic.py\", line 2002, in atomic_move\nUnboundLocalError: local variable 'in_src' referenced before assignment\n", "module_stdout": "", "msg": "MODULE FAILURE"}
        to retry, use: --limit @/tmp/ansible/etc_host.retry
```

This is an error raised by the incorrect handling of file not found exception. Removing this line handler (line 2002) we can get to the actual issue:

```
TASK [insert 127.0.0.2 line in /etc/hosts] *************************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Could not write data to file (/etc/hosts) from (/tmp/tmpll43Kj): [Errno 2] No such file or directory: '/tmp/tmpll43Kj'"}
        to retry, use: --limit @/tmp/ansible/etc_host.retry
```

pgrange replied:
OK, now it is clearer. I am sure I want b_tmp_dest_name and not b_src because the exception is raised by os.rename(b_tmp_dest_name, b_dest) line 1987.

So when we catch the exception and try to unsafe_write, we just failed to os-rename b_tmp_dest_name so we must unsafe_write b_tmp_dest_name

bcoca replied:
no, the exception is around this code:
`os.rename(b_src, b_dest)`

as the original exception on this 'downward path', checking if we are improperly resetting b_src between that and this line

bcoca replied:
so this looks like a problem `shutil.move(b_src, b_tmp_dest_name)`

bcoca replied:
but it can also fail here `shutil.copy2(b_src, b_tmp_dest_name)`, so we might just need a flag on 'which to use'

abadger replied:
You could either do an os.path.exists check here to decide based on whether b_src exists or you could abstract the unsafe_writes fallback into a helper method, and then write two try: excepts that target each section of code.  If the code fails in  shutil.copy2 pr shutil.move then call the helper function with b_src.  If it fails in os.rename then call the helper with b_tmp_dest_name.  atomic_move is getting rather long so the latter may be the better option.

Commit:
Abstract the unsafe_writes fallback into a helper method.
Explicitly try/except os.rename part of the code and call this helper method.
If the code fails in shutil.copy2 or shutil.move this should not be related to issue #13981
since they write to b_tmp_dest_name.
(as suggested by @abadger)

pgrange commented:
@abadger I extracted unsafe_writes code as suggested.
I focused the try/except code that relies on _unsafe_writes() around os.rename code.
However I did not try/except around shutil.copy2 and shutil.move since they are writing to b_tmp_dest_name and no problem can occur when writing to b_tmp_dest_name... no problem that _unsafe_writes can solve, actually.
See #13981 for details about the problem that _unsafe_writes tries to solve.

Regards,

abadger commented on lib/ansible/module_utils/basic.py:
```diff
@@ -2022,6 +2005,28 @@ def atomic_move(self, src, dest, unsafe_writes=False):
+    def _unsafe_writes(self, unsafe_writes, src, dest, exception):
+      # sadly there are some situations where we cannot ensure atomicity, but only if
+      # the user insists and we get the appropriate error we update the file unsafely
+      if unsafe_writes and exception.errno == errno.EBUSY:
```
I'd do this checking in the caller.  That way the function call reads as "Do an unsafe write".  As it's written now it's more of a "I think we should do an unsafe_write.  Do you agree?"

abadger commented:
Only one comment.  Rest looks good to me.  I'll give @bcoca a chance to reply in case he sees something else.

Commit:
Check if unsafe_writes in the caller, not in _unsafe_writes.
That way the function call reads as "Do an unsafe write"
and not as "I think we should do an unsafe_write.

pgrange commented:
@abadger agree. New commit checks if unsafe_writes in the caller.

bcoca commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1983,29 +1983,15 @@ def atomic_move(self, src, dest, unsafe_writes=False):
                             if e.errno != errno.EPERM:
                                 raise
-                        os.rename(b_tmp_dest_name, b_dest)
+                        try:
```
so this changes the point of coverage of unsafe_writes, it will not be used as fallback except for the narrower error on temp file renaming.

abadger replied:
The answer to that is in the commit message: "If the code fails in shutil.copy2 or shutil.move this should not be related to issue #13981 since they write to b_tmp_dest_name."

pgrange commented:
Great. Thanks !
